File: blind/UniquePaths.js
/**
 * A robot is located on the top left corner of a m x n grid. It can only move down or right. How many possible unique paths are there?
 * 
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
// Top Down Recursion
const uniquePaths = (m, n) => {
	// edge case
	if (m === 1 || n === 1) return 1;

	memo = new Map();

	// not counting index 0
	const helper = (row, column) => {
		// base case
		if (row === m && column === n) return 1;

		// failure case
		if (row > m || column > n) return 0;

		let key = row.toString() + ',' + column.toString();

		// console.log(key, memo)

		if (memo.has(key)) {
			return memo.get(key);
		}

		let result = helper(row + 1, column) + helper(row, column + 1);

		memo.set(key, result);

		return result;
	};

	return helper(1, 1);
};

// Bottom Up Dynamic Programming
const uniquePaths = (m, n) => {
	// edge case
	if (m === 1 || n === 1) return 1;

	// create a m + 1 x n + 1 matrix dp to represent board
	const inner = new Array(n + 1).fill(0);
	const dp = [];
	for (let i = 0; i < m + 1; i++) dp.push([...inner]);
	console.log(dp);

	// dp[index][index] represents unique paths if it is the finish point
	dp[1][1] = 1; 

	// console.log(dp);

	// a particular place is calculated by adding the unique paths results from top and left
	for (let i = 1; i < dp.length; i++) {
		for (let j = 1; j < dp[0].length; j++) {
			dp[i][j] = dp[i][j] + dp[i - 1][j] + dp[i][j - 1];
			// console.log(dp);
		}
	}

	// console.log(dp);

	return dp[dp.length - 1][dp[0].length - 1];
};

File: golang/unique_paths.go
func uniquePaths(m int, n int) int {
	paths := 0
	helper(m, n, 1, 1, &paths)
	return paths
}

// times out
func helper(m int, n int, y int, x int, paths *int) {
	// base case
	if y == m && x == n {
		(*paths)++
		return
	}

	// avoid bound issues
	if y != m {
		helper(m, n, y+1, x, paths)
	}
	if x != n {
		helper(m, n, y, x+1, paths)
	}
}

// dynamic programming
func uniquePaths(m int, n int) int {
	// the paths for any arbitary spot is the sum of paths for x-1 and y-1 if not on bound
	dp := make([][]int, m)
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n)
	}

	// init
	for row := 0; row < m; row++ {
		dp[row][0] = 1
	}
	for col := 0; col < n; col++ {
		dp[0][col] = 1
	}

	// sum left and up spots
	for row := 1; row < m; row++ {
		for col := 1; col < n; col++ {
			dp[row][col] = dp[row][col-1] + dp[row-1][col]
		}
	}

	return dp[m-1][n-1]
}